src/lib/modules/scaffolding/index.ts

```diff
@@ -1,63 +1,10 @@
-import { Contract } from "../../../typings/contract";
 import { Embark } from "../../../typings/embark";
-import { CommandOptions, ContractLanguage, Framework } from "./commandOptions";
-import { SolidityBuilder } from "./contractLanguage/solidityBuilder";
-import { ReactBuilder } from "./framework/reactBuilder";
-import { SmartContractsRecipe } from "./smartContractsRecipe";
+import { GenerateContractCommand } from "./generateContractCommand";
+import { GenerateUiCommand } from "./generateUiCommand";
 
 export default class Scaffolding {
-
-  constructor(private embark: Embark, private options: any) {
-    this.embark.events.setCommandHandler("scaffolding:generate:contract", (cmdLineOptions: any,  cb: (files: string[]) => void) => {
-      this.generateContract(cmdLineOptions).then(cb);
-    });
-
-    this.embark.events.setCommandHandler("scaffolding:generate:ui", (cmdLineOptions: any,  cb: (files: string[]) => void) => {
-      this.generateUi(cmdLineOptions).then(cb);
-    });
-  }
-
-  private contractLanguageStrategy(recipe: SmartContractsRecipe, options: CommandOptions) {
-    switch (options.contractLanguage) {
-      case ContractLanguage.Solidity: {
-        return new SolidityBuilder(this.embark, recipe, options);
-      }
-    }
-  }
-
-  private frameworkStrategy(recipe: SmartContractsRecipe, contracts: Contract[], options: CommandOptions) {
-    switch (options.framework) {
-      case Framework.React: {
-        return new ReactBuilder(this.embark, recipe, contracts, options);
-      }
-    }
-  }
-
-  private parseAndValidate(cmdLineOptions: any) {
-    const options = new CommandOptions(this.embark.logger, cmdLineOptions.framework, cmdLineOptions.contractLanguage, cmdLineOptions.overwrite);
-    const recipe = new SmartContractsRecipe(this.embark.logger, cmdLineOptions.contractOrFile, cmdLineOptions.fields);
-    options.validate();
-    recipe.validate();
-
-    return {recipe, options};
-  }
-
-  private async generateContract(cmdLineOptions: any) {
-    const {recipe, options} = this.parseAndValidate(cmdLineOptions);
-    return await this.contractLanguageStrategy(recipe, options).build();
-  }
-
-  private async generateUi(cmdLineOptions: any) {
-    const {recipe, options} = this.parseAndValidate(cmdLineOptions);
-    const contracts = await this.getContracts();
-    return await this.frameworkStrategy(recipe, contracts, options).build();
-  }
-
-  private getContracts() {
-    return new Promise<Contract[]>((resolve) => {
-      this.embark.events.request("contracts:list", (_: null, contracts: Contract[]) => {
-        resolve(contracts);
-      });
-    });
+  constructor(embark: Embark, options: any) {
+    new GenerateContractCommand(embark).observe(embark.events);
+    new GenerateUiCommand(embark).observe(embark.events);
```

Comment:
So this is where it gets "critical".

As you can see we don't have a way to actually do anything with the values that are being emitted here (in this case we probably aren't interested in it anyways), but it basically makes it non-lazy by default, which takes away a main characteristic of observables.

Also, API-wise, I think it'd be nicer if we could use our EventBus to register a new command, not the other way around.

E.g.

```
embark.events.registerCommand(GenerateUICommand, new GenerateUICommand(embark));
embark.events.registerCommand(GenerateContractCommand, new GenerateContractCommand(embark));

embark.events.observe(GenerateUICommand).pipe(
  map(val => /* do something with val */)
).subscribe(val => {
  // ...
});
embark.events.observe(GenerateContractCommand).subscribe();
```

Something along those lines.

Reply:
yes indeed, this is why I asked for review early, the API is the critical part.
Also something we haven't discusses is how we request event. That forces us to have the type any in all our handler, quite annoying.

After sleeping on it I was considering having this API:

`events.request(Commands.generateContract())` where generate contracts contains the key (a string or symbol), and the params

```
embark.events.registerCommand
```

=> in the modules they are not command but handler, so I would do:
```
embark.events.observe(Commands.GENERATE_CONTRACT).subscribe(new GenerateContractHandler())
```

What do you think?

Also if you look carefully what we are doing is just rebuilding redux pattern